### PR TITLE
Extend ioext multi dev

### PR DESCRIPTION
**Pull request by andreaskuster**

### CODE REVIEW

**Comment by andreaskuster** [comment]
@KarlSchu I deleted your non-generic code in order to extend this class. Please make sure you grab the necessary lines from the diff and include it.

File: include/definitions.h [hunk 1]
```diff
@@ -36,8 +36,11 @@
 #define OLED_HEIGHT 64
 #define I2C_ADDRESS_SSD1305 0x3C
 
+#define PCF8574_NUM_DEVICES 3
 #define PCF8574_NUM_PORTS 8
-#define I2C_ADDRESS_PCF8574_IOExt2 0x20
+#define I2C_ADDRESS_PCF8574_IOExt0 0x20 // TODO: add correct address!!!
+#define I2C_ADDRESS_PCF8574_IOExt1 0x20 // TODO: add correct address!!!
+#define I2C_ADDRESS_PCF8574_IOExt2 0x20 // TODO: add correct address!!!
```
Comment (KarlSchu) [inline include/definitions.h:43]: #define I2C_ADDRESS_PCF8574_IOExt3 0x20 // TODO: add correct address!!!

File: lib/IOExt/IOExt.h [hunk 2]
```diff
@@ -1,33 +1,65 @@
...
 class IOExt : public abstract_task {
 private:
-  // TODO: is there an IOExt1? If yes: we should integrate this here too, see ADC multi device as a reference, Else: remove '2'
-  PCF8574 IOExt2 = PCF8574(I2C_ADDRESS_PCF8574_IOExt2, I2C_SDA, I2C_SCL, I2C_INTERRUPT_PIN_PCF8574, keyPressedInterruptHandler);
+
+  PCF8574 IOExt[3] = {
+    PCF8574(I2C_ADDRESS_PCF8574_IOExt0, I2C_SDA, I2C_SCL, I2C_INTERRUPT_PIN_PCF8574, keyPressedInterruptHandler),
+    PCF8574(I2C_ADDRESS_PCF8574_IOExt1, I2C_SDA, I2C_SCL, I2C_INTERRUPT_PIN_PCF8574, keyPressedInterruptHandler),
+    PCF8574(I2C_ADDRESS_PCF8574_IOExt2, I2C_SDA, I2C_SCL, I2C_INTERRUPT_PIN_PCF8574, keyPressedInterruptHandler)
+  };
```
Comment (KarlSchu) [inline lib/IOExt/IOExt.h:19]: ,
PCF8574(I2C_ADDRESS_PCF8574_IOExt3, I2C_SDA, I2C_SCL, I2C_INTERRUPT_PIN_PCF8574, keyPressedInterruptHandler)

Comment (KarlSchu) [reply lib/IOExt/IOExt.h:19]: we have 4 IOExt 
... and all resulting code too

File: lib/IOExt/IOExt.cpp [hunk 3]
```diff
@@ -25,22 +19,10 @@ void IOExt::init() {
...
-  // start
-  if (IOExt2.begin()) {
-    printf("[v] IOExt2 inited.\n");
-  } else {
-    printf("[x] IOExt2 init failed.\n");
-    // TODO: action for init error?
+  for(int i = 0; i < PCF8574_NUM_DEVICES; i++){
+    if(IOExt[i].begin()){
+      printf("Initialization of IOExt0%u failed.\n", i);
+    }
```
Comment (KarlSchu) [inline lib/IOExt/IOExt.cpp:25]: This format is for the boot screen importend to see what works and what fails:
  printf("[v] IOExt2 inited.\n");
  } else {
    printf("[x] IOExtXXX init failed.\n");
    // TODO: action for init error?

File: lib/IOExt/IOExt.cpp [hunk 4]
```diff
@@ -54,91 +36,47 @@ void IOExt::exit(void) {
...
-  PCF8574::DigitalInput dra = IOExt2.digitalReadAll();
-  xSemaphoreGive(i2cBus.mutex);
 
-  if (dra.p0 & dra.p1 & dra.p2 & dra.p3 & dra.p4 & dra.p5 & dra.p6 & dra.p7) {
-    taskSleep = 50; // default value for fast reaction to pressed button
-    return;
+  // read all
+  PCF8574::DigitalInput dra[PCF8574_NUM_DEVICES];
+  for(int i = 0; i < PCF8574_NUM_DEVICES; i++){
+    dra[i] = IOExt[i].digitalReadAll();
```
Comment (KarlSchu) [inline lib/IOExt/IOExt.cpp:43]: digitalReadAll() doesn't work as expected

**Comment by andreaskuster** [comment]
@andreaskuster 
- [ ] Add dummy function for the deleted code

File: include/definitions.h (hunk 1 above)
Comment (andreaskuster) [reply include/definitions.h:43]: Looking at the datasheet, the configurable address range is from `0x20` to `0x28`. Setting addresses to `0x20, 0x21, 0x22` and changing the todo to: `TODO: set jumpers accordingly` 

grafik

File: lib/IOExt/IOExt.h (hunk 2 above)
Comment (andreaskuster) [reply lib/IOExt/IOExt.h:19]: Thanks for the hint.

File: lib/IOExt/IOExt.cpp (hunk 3 above)
Comment (andreaskuster) [reply lib/IOExt/IOExt.cpp:25]: Good point, has been fixed.

File: lib/IOExt/IOExt.cpp (hunk 4 above)
Comment (andreaskuster) [reply lib/IOExt/IOExt.cpp:43]: By defining `USE_LEGACY_PIN_READ`, i added single pin read functionaliy

**Review by KarlSchu — APPROVED** [review]

**Pull request merged**